server/src/main/java/io/spine/server/delivery/GroupByTargetAndDeliver.java

```diff
@@ -58,12 +56,11 @@
      */
     @Override
     public DeliveryErrors executeFor(List<InboxMessage> messages) {
-        Map<String, List<InboxMessage>> messagesByType = groupByTargetType(messages);
-
+        List<Segment> segments = groupByTargetType(messages);
         DeliveryErrors.Builder errors = DeliveryErrors.newBuilder();
-        for (String typeUrl : messagesByType.keySet()) {
-            ShardedMessageDelivery<InboxMessage> delivery = inboxDeliveries.get(typeUrl);
-            List<InboxMessage> deliveryPackage = messagesByType.get(typeUrl);
+        for (Segment segment : segments) {
```

Comment:
> Can we run this in parallel (i.e. via segments.parallelStream()?

Unfortunately, no. As we must deliver the messages exactly in the order, in which they are listed. 

Different segments are produced wisely, so that across the list of segments all the messages are ordered by their time-of-emission. In this way we guarantee that two signals will be dispatched — no matter to which destination — exactly in the order they were emitted. 

Also, different entities to which the messages are dispatched, may read one another. So even from this point of view, we must not update one entity earlier than it is expected to be updated, since we may break the end-user logic.

Reply:
I have extracted and improved some bits here and there, including switching to `ImmutableList` when producing segments.

However, the `Segment`-grouping code is hardly simplify-able (I tried!) with the same amount of data structures involved. And we need to keep up the performance here, as the delivery should be as fast as possible. So I refrained from modifying it as such.